### PR TITLE
[On Policy Training] Multi-Turn Response Regeneration

### PR DESCRIPTION
**Pull request by WindChimeRan**
## Purpose

Response regeneration should support multi-turn (currently single-turn only). The driver is exposure bias: off-policy multi-turn data conditions each turn on the dataset's responses rather than the target's, so every turn past the first is misaligned. Regenerating turn-by-turn against the on-policy prefix fixes this — and yields longer, more realistic context as a bonus.

## Summary

- Multi-turn output. For any dataset with a messages/conversations field (ultrachat, magpie, open-perfectblend…), every user turn is regenerated against the freshly generated prefix, so a conversations list can now be many turns long instead of always exactly [human, gpt].
- System prompt preserved. A source system turn now appears in the output ({"from":"system",…}) and is used as context during regeneration. Previously system was always dropped.
- Per-turn reasoning_content. Reasoning is now attached to each assistant turn inside conversations. Previously it was a single field in top-level metadata (which the training pipeline drops).

## New behavior

- New <outfile>.errors.jsonl → partial multi-turn records were otherwise indistinguishable from clean ones downstream. (align with SpecForge)
- Empty/None assistant content → treated as a failed turn

## Out-of-scope

* tool use
* multi-modal
* concurrency
* doc: This PR indeed expands the behavior a bit. But I plan to make a larger refactoring PR to the doc, to make response regeneration explicitly the default. 

## Tests

* new unit tests, with diverse dataset format preset
* verified end2end: Qwen2.5-0.5B + `--dataset ultrachat --limit 2`

## Checklist

I have filled in:

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan/results, such as providing test command and pasting the results.
- [ ] (Optional) The necessary documentation update.
- [x] I (a human) have written or reviewed the code in this pr to the best of my ability.

### CODE REVIEW

**Review by orestis-z — COMMENTED** [review]
Taking this one to review

File: scripts/response_regeneration/script.py [hunk 1]
```diff
@@ -154,74 +191,107 @@ async def worker(
...
+                    data = await response.json()
+
+                choice = data["choices"][0]
+                message = choice["message"]
+                generated_text = message.get("content")
+                reasoning_content = message.get("reasoning_content")
+                if reasoning_content is None:
+                    reasoning_content = message.get("reasoning")
+                finish_reason = choice.get("finish_reason")
+                usage = data.get("usage")
```
Comment (orestis-z) [inline scripts/response_regeneration/script.py]: [minor] `usage` only captures the last turn. For multi-turn conversations with N user turns, N API calls are made, but `usage` in the output metadata only reflects the final call. Total prompt+completion tokens across the conversation would be more useful for cost tracking. Consider accumulating:

```python
total_usage = {"prompt_tokens": 0, "completion_tokens": 0, "total_tokens": 0}
```

and summing each turn's usage into it.

File: scripts/response_regeneration/script.py [hunk 2]
```diff
@@ -154,74 +191,107 @@ async def worker(
...
+                async with sem, session.post(endpoint, json=payload) as response:
+                    data = await response.json()
+
+                choice = data["choices"][0]
+                message = choice["message"]
+                generated_text = message.get("content")
+                reasoning_content = message.get("reasoning_content")
+                if reasoning_content is None:
+                    reasoning_content = message.get("reasoning")
+                finish_reason = choice.get("finish_reason")
```
Comment (orestis-z) [inline scripts/response_regeneration/script.py]: [minor] `finish_reason` only captures the last turn. If an earlier turn hit `length` but the last was `stop`, the metadata would hide the truncation. A `finish_reasons` list (or at least flagging if any turn was non-`stop`) would be more informative.

File: scripts/response_regeneration/script.py [hunk 3]
```diff
@@ -320,8 +397,8 @@ async def main():
                 if args.language_filter and row.get("language") != args.language_filter:
                     continue
 
-                prompt = row.get(prompt_field)
-                if not prompt:
+                turns = extract_turns(row, prompt_field)
+                if not any(turn["role"] == "user" for turn in turns):
```
Comment (orestis-z) [inline scripts/response_regeneration/script.py]: [nit] Redundant check — `extract_turns` already guarantees any non-empty return contains at least one user turn (line 134). So this is equivalent to `if not turns:` but scans the list.

File: scripts/response_regeneration/script.py [hunk 4]
```diff
@@ -262,10 +332,15 @@ async def main():
...
+    error_outfile = f"{base}.errors{ext or '.jsonl'}"
+
     print(f"Using dataset: {dataset_id}")
     print(f"Split: {split}")
     print(f"Prompt field: {prompt_field}")
     print(f"Output file: {args.outfile}")
+    print(f"Error file: {error_outfile}")
     print()
 
     seen_ids = load_seen(args.outfile) if args.resume else set()
```
Comment (orestis-z) [inline scripts/response_regeneration/script.py:362]: [minor] `--resume` will silently retry previously-failed rows since `load_seen` only reads the success file, not `*.errors.jsonl`. This is probably the right behavior (transient endpoint errors should be retried), but worth a one-line comment or `--help` note — a user seeing a row in the error file and then seeing it re-processed on resume might be confused.

File: scripts/response_regeneration/script.py [hunk 5]
```diff
@@ -154,74 +191,107 @@ async def worker(
...
+
+                prefix.append({"role": "user", "content": turn["content"]})
+                out_convs.append({"from": "human", "value": turn["content"]})
+
+                payload = {
+                    "model": args.model,
+                    "messages": prefix,
+                    "max_tokens": args.max_tokens,
+                }
+                async with sem, session.post(endpoint, json=payload) as response:
```
Comment (orestis-z) [inline scripts/response_regeneration/script.py]: [minor] Semaphore is acquired per-turn, not per-conversation. Each turn independently acquires the semaphore, so request concurrency stays bounded (good). But a single 5-turn conversation occupies a worker for 5 sequential acquisitions while other workers may idle waiting. With high-turn-count datasets this could reduce throughput vs. one-request-per-item. Not a problem for current datasets but worth noting.

**Review by orestis-z — COMMENTED** [review]
Thanks for the contribution @WindChimeRan — this is a well-motivated and cleanly implemented extension. The on-policy multi-turn regeneration is a meaningful improvement, and the test coverage (especially the downstream normalizer contract test) is solid. A few notes below, nothing blocking.

**Review by orestis-z — APPROVED** [review]

**Review by orestis-z — COMMENTED** [review]
Non-blocking: multi-turn e2e smoke test (even just 1 sample from ultrachat with a real or mocked server) in CI but not blocking given this lives in scripts/, not core library code.

File: scripts/response_regeneration/script.py (hunk 4 above)
Comment (WindChimeRan) [reply scripts/response_regeneration/script.py:362]: I think there are some pre-existing bugs before this PR. Will try to fix it in followup PR.

File: scripts/response_regeneration/script.py (hunk 5 above)
Comment (WindChimeRan) [reply scripts/response_regeneration/script.py]: will address semaphore in a dedicated followup PR. (pre-existing issue, code is not clean)

**Comment by WindChimeRan** [comment]
E2E ablation study last week for multi-turn on-policy training: 

### DSpark (Qwen3-8B): training-data policy vs. validation set

Our speculator train & eval is default on **off-policy** (not response re-generation). This is not ideal

| Training responses              | Validation set | full_acc | pos1  | pos4  | pos7  |
|---------------------------------|----------------|----------|-------|-------|-------|
| off-policy (ShareGPT)           | on-policy      | 0.280    | 0.550 | 0.228 | 0.156 |
| on-policy, single-turn          | on-policy      | 0.374    | 0.609 | 0.334 | 0.250 |
| on-policy, multi-turn   #632        | on-policy      | 0.405    | 0.648 | 0.365 | 0.272 |

The bottom three rows share the same held-out on-policy validation set, so they are directly comparable.

**Comment by WindChimeRan** [comment]
Followup PR soon
* semaphore cleanup
* investigate into `--resume`, could be a bug. 
* sample a mixed small set for CI testing
* doc update. multi-turn on policy should be the default.

**Comment by shanjiaz** [comment]
@WindChimeRan @orestis-z  I think @imargulis  has mentioned the unreliable resume bug in this [RFC](https://github.com/vllm-project/speculators/issues/584) and is working on the refactor. Agree not blocking for this PR.

**Review by shanjiaz — APPROVED** [review]
I think the logic is sound, but would prefer we cleanup the comments a bit before merging. Thanks

**Comment by benchislett** [comment]
@WindChimeRan can you compare against a _test_ set here? For example, our SPEED-Bench has several multi-turn conversations that are designed for "safe" reuse where collisions are not likely. The "turns" column specifically has multiple prompts you can send in sequence to measure AL over multiple turns without worrying about incoherent decay. PTAL

https://huggingface.co/datasets/nvidia/SPEED-Bench

**Comment by WindChimeRan** [comment]
Thanks @benchislett 

I eval it on SPEED-Bench multi-turn conversations (120 convs with ≥2 turns), sending each turns entry in sequence and feeding the model's own reply back as context. (note that this is a out-of-distribution eval. The speculator is only trained on SharedGPT)

Training data | AL overall | turn 1 | turn 2 | turn 3+
-- | -- | -- | -- | --
on-policy single-turn | 2.031 | 2.019 | 2.066 | 1.628
on-policy multi-turn (this PR) | 2.169 | 2.153 | 2.199 | 1.848
